font/src/directwrite/mod.rs

```diff
@@ -0,0 +1,222 @@
+// Copyright 2019 Joe Wilm, The Alacritty Project Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//! Rasterization powered by DirectWrite
+extern crate dwrote;
+use self::dwrote::{
+    FontCollection, FontStretch, FontStyle, FontWeight, GlyphOffset, GlyphRunAnalysis,
+};
+
+use super::{FontDesc, FontKey, GlyphKey, Metrics, RasterizedGlyph, Size, Slant, Style, Weight};
+
+pub struct DirectWriteRasterizer {
+    fonts: Vec<dwrote::FontFace>,
+    device_pixel_ratio: f32,
+}
+
+impl crate::Rasterize for DirectWriteRasterizer {
+    type Err = Error;
+
+    fn new(device_pixel_ratio: f32, _: bool) -> Result<DirectWriteRasterizer, Error> {
+        Ok(DirectWriteRasterizer { fonts: Vec::new(), device_pixel_ratio })
+    }
+
+    fn metrics(&self, key: FontKey, size: Size) -> Result<Metrics, Error> {
+        let font = self.fonts.get(key.token as usize).ok_or(Error::FontNotLoaded)?;
+
+        let vmetrics = font.metrics();
+        let scale = (size.as_f32_pts() * self.device_pixel_ratio * (96.0 / 72.0))
+            / f32::from(vmetrics.designUnitsPerEm);
+
+        let underline_position = f32::from(vmetrics.underlinePosition) * scale;
+        let underline_thickness = f32::from(vmetrics.underlineThickness) * scale;
+
+        let strikeout_position = f32::from(vmetrics.strikethroughPosition) * scale;
+        let strikeout_thickness = f32::from(vmetrics.strikethroughThickness) * scale;
+
+        let line_height = f64::from(
+            i32::from(vmetrics.ascent) + i32::from(vmetrics.descent) + i32::from(vmetrics.lineGap),
```

Comment:
There are a lot of conversions here, could we instead use one of the types of the metrics to convert to so we do not need to convert everything?

Reply:
I think this should now be fixed - let me know if this isn't what you meant.

Reply:
That was actually not exactly what I was proposing.

Ascent and descent are both u16 while the line gap is i16. So to reduce conversions as much as possible, it should be possible to add all of them together by just converting the ascent and descent instead of having to convert everything.